qadence/noise/protocols.py

```diff
@@ -33,27 +66,161 @@ def get_noise_fn(self) -> Callable:
         return cast(Callable, fn)
 
     def _to_dict(self) -> dict:
-        return {"protocol": self.protocol, "options": self.options}
+        return {
+            "protocol": self.protocol,
+            "options": self.options,
+            "protocol_type": self.protocol_type,
+        }
 
     @classmethod
-    def _from_dict(cls, d: dict) -> Noise | None:
+    def _from_dict(cls, d: dict | None) -> NoiseSource | None:
         if d:
-            return cls(d["protocol"], **d["options"])
+            protocol_type = d.get("protocol_type", "")
+            return cls(d["protocol"], **d["options"], protocol_type=protocol_type)
         return None
 
     @classmethod
     def list(cls) -> list:
         return list(filter(lambda el: not el.startswith("__"), dir(cls)))
 
+    def __repr__(self) -> str:
+        return f"{self.protocol_type} NoiseSource({self.protocol}, {str(self.options)})"
+
+
+class NoiseHandler:
+    """A container for multiple sources of noise."""
+
+    def __init__(
+        self,
+        protocol: str | NoiseSource | list[str] | list[NoiseSource],
```

Comment:
Do we still need `str` ?

Reply:
Not anymore with the new definition of NoiseProtocol